pkg/plugin/config.go

```diff
@@ -226,10 +226,10 @@ func (c *nodeConfig) vCpuLimits(total *resource.Quantity) (_ nodeResourceState[u
 	margin = resource.NewMilliQuantity(unreservableCpuMillis, c.Cpu.System.Format)
 	margin.Sub(c.Cpu.System)
 
-	return nodeResourceState[uint16]{
-		Total:                uint16(totalRounded),
-		System:               uint16(systemCpus),
-		Watermark:            uint16(c.Cpu.Watermark * float32(reservableCpus)),
+	return nodeResourceState[api.MilliCPU]{
+		Total:                api.MilliCPU(totalRounded * 1000),
+		System:               api.MilliCPU(systemCpus * 1000),
+		Watermark:            api.MilliCPU(c.Cpu.Watermark * float32(reservableCpus) * 1000),
```

Comment:
There's definitely a more thorough to make this better (i.e. don't count "full" CPUs), but realistically that's part of a broader change to the scheduler to move away from counting everything in terms of compute units.

So: this is going to add a bit of scheduling inefficiency, but that's ok for now I think. Just want to flag this for later.

Reply:
Could you elaborate what's the deficiency here (out of curiosity)?

Reply:
(note: this is defined by `vCpuLimits` ([pkg/plugin/config.go](https://github.com/neondatabase/autoscaling/blob/b206290f5898f263a79682fb08de9d1ec485b503/pkg/plugin/config.go#L206-L237)) and `buildInitialNodeState` ([pkg/plugin/state.go](https://github.com/neondatabase/autoscaling/blob/b206290f5898f263a79682fb08de9d1ec485b503/pkg/plugin/state.go#L476-L553)) which calls it)

Sure: As _currently_ written (again: it's bad, we should change it) - the scheduler determines the allocatable CPU as follows:

1. Set `total := Node.Status.Allocatable().Cpu()` (or `.Capacity().Cpu()` if not present) — provided by the K8s API
2. Subtract "system" CPUs from `total` (amount of CPU reserved for daemon pods, etc)
3. Round down `total` to the nearest integer

So basically we're rounding `total` down, which is now not needed. This is not _really_ clear in `vCpuLimits`, unfortunately, because there's some other things it's trying to do as well.

There may also be some issues with how `otherPodResources` interacts with this - going to double-check that during my next review.

Reply:
Double-checked: Looks like this diff made it ok, because it's no longer rounding:

```diff
diff --git a/pkg/plugin/state.go b/pkg/plugin/state.go
index dc5d508..f2ef947 100644
--- a/pkg/plugin/state.go
+++ b/pkg/plugin/state.go
@@ -301,8 +301,7 @@ func (r *nodeOtherResourceState) calculateReserved(memSlotSize *resource.Quantit
 		// set cupCopy := r.rawCpu - r.marginCpu
 		cpuCopy := r.RawCPU.DeepCopy()
 		cpuCopy.Sub(*r.MarginCPU)
-		// note: Value() rounds up, which is the behavior we want here.
-		r.ReservedCPU = uint16(cpuCopy.Value())
+		r.ReservedCPU = api.MilliCPUFromResourceQuantity(cpuCopy)
 	}
 
 	// If rawMemory doesn't exceed the margin ..., set reserved = 0
```

so no issues :)